common/sphinx.c

```diff
@@ -669,3 +785,98 @@ u8 *unwrap_onionreply(const tal_t *ctx,
 		return tal_free(final);
 	return final;
 }
+
+struct onionpacket *
+sphinx_decompress(const tal_t *ctx,
+		     const struct sphinx_compressed_onion *src,
+		     struct secret *shared_secret)
+{
+	struct onionpacket *res = tal(ctx, struct onionpacket);
+	size_t srclen = tal_bytelen(src->routinginfo);
+	size_t prefill_size = ROUTING_INFO_SIZE - srclen;
+
+	res->version = src->version;
+	res->ephemeralkey = src->ephemeralkey;
+	memcpy(res->mac, src->mac, HMAC_SIZE);
+
+	/* Decompress routinginfo by copying the unmodified prefix, setting
+	 * the compressed suffix to 0x00 bytes and then xoring the obfuscation
+	 * stream in place. */
+	memset(res->routinginfo, 0, ROUTING_INFO_SIZE);
+	memcpy(res->routinginfo, src->routinginfo, srclen);
+	sphinx_prefill_stream_xor(res->routinginfo + srclen, prefill_size,
+				  shared_secret);
+
+	return res;
+}
+
+struct sphinx_compressed_onion *
+sphinx_compress(const tal_t *ctx, const struct onionpacket *packet,
+		struct sphinx_path *path)
+{
+	struct sphinx_compressed_onion *res;
+	size_t payloads_size = sphinx_path_payloads_size(path);
+
+	/* We can't compress an onion that doesn't have a rendez-vous node. */
+	if (path->rendezvous_id)
+		return NULL;
```

Comment:
This is clearly untested? :(

Reply:
Right, I replaced the `serialize_compressed_onion` method with the compress + serialize primitives I added. I also added a check to make sure non-compressible onions return `NULL`.